### ISSUE TITLE
ZestFinance writeup on SHAP and why it shouldn't be used on its own

### ISSUE THREAD

**Issue by InfiniteDoLoop** [shap/shap#624]
Hi Scott,

First off, thanks for the work you've done on SHAP. I think it'll definitely help people get more comfortable with ML model results and make them mainstream.

I came across the following article from **ZestFinance** which references your package and was an interesting read:
[Why Lenders Shouldn't 'Just Use SHAP' To Explain Machine Learning Credit Models](https://www.zestfinance.com/blog/why-you-shouldnt-just-use-shap-to-power-your-ml-explainability-requirements)

It seems to have some details on what your package is missing and what they've enhanced to make it usable for credit underwriting. Wanted to check in with you for some clarification as it seems this package does include some of these functions they describe but I could be wrong. Apologies in advance for the barrage of questions btw :)

**1) Margin Space vs Score Space**

> The score space is very different mathematically from the credit model’s actual output, which is said to be in “margin space.” Margin space numbers fall in a narrow range from 0 to 1. In general, the relationship between the model’s actual output in margin space and the acceptance threshold in score space is extremely non-linear, and you have to transform the model’s output to generate the number the lending business wants.

It sounds like **Margin Space** refers to raw output from obtained from XGboost model in log odds scale and **Score Space** is this output transformed to probability space since they mention it is a non-linear transformation. Is this something currently supported when running:

`shap.TreeExplainer(model, model_output = 'probability', feature_dependence = "independent", data = X)`

**2) Explanation by reference**

> There are many details you need to get right in this process, including the appropriate application of sample weights, mapping to score space at the approval cut-off, sampling methods, and accompanying documentation. Out of the box, SHAP doesn’t allow you to easily do this.

I think the previous code in *(1)* is still applicable here, but the sampling of X is important to obtain an appropriate reference point. As an example, let 80% be the cutoff, where a score below 80% is approved for credit. Should the subsample of X include all potential approvals or should it be sampled such that the expected value is equal to 80% to obtain a valid reference point?

**3) You want to use modeling methods other than XGBoost**

> What’s more, SHAP cannot explain ensembles of continuous and tree-based models, such as stacked or deeply stacked models that combine xgboost and deep neural networks.

I'm sure the other model types can be tackled by the explainers included in your package. When it comes to stacking, the shap values for each variable in the stacked model can be determined by computing a weighted sum of the shap values from the individual models. Is this correct?

**4) Even on a simple XGBoost model, SHAP fails to uncover the underlying geometry** 

I don't have a specific question on this section, but would be interested in any comments/insights you may have on why the separation of SHAP values is less distinct than what ZAML produces. 

Thanks for your time and looking forward to seeing what you'll be working on next!

**Comment by slundberg** [comment]
Great questions! I know Jay and assume he has done good work there at Zest. Some of the issues they address do seem to be supported in our new releases. I’ll comment in more detail soon. Sorry the slowness lately, I defend my dissertation on Monday so I have had my head down for a bit :)

**Comment by InfiniteDoLoop** [comment]
> I’ll comment in more detail soon. Sorry the slowness lately, I defend my dissertation on Monday so I have had my head down for a bit :)

Of course, good luck on Monday!

**Comment by jphall663** [comment]
@nickpschmidt - FYI ^

**Comment by firmai** [comment]
@slundberg good luck with your dissertation.

**Comment by slundberg** [comment]
Below are my thoughts on the recent Zest article *"Why Not Just Use SHAP?"*. I took a while to respond here because I wanted to first chat with Jay and John at Zest who have worked on Zest's explainability development. I think both Jay and John have a solid understanding of both finance and explainability and are doing good things. Unfortunately, I think this particular article was influenced too much by "marketing pressures" that encourage Zest to position itself as the go-to for explainable finance.

Below are specific responses to quoted sections from the original article that require correction or clarification:


**1) Margin Space vs Score Space**

> Margin space numbers fall in a narrow range from 0 to 1

This is not true for gradient boosted decision trees such as XGBoost, where the margin can take many different meanings (units) depending on the loss function that is used. Assuming we are doing binary classification, the "margin" is the value before the logistic link function and can take arbitrary positive or negative values (it is in log odds units). The default output for XGBoost and LightGBM models explained by shap is this "true" margin value.


> The problem with SHAP is that, because of the way it computes its Shapley values, it really only works in margin space.

The exact solution to the Shapley value equation used by shap.TreeExplainer is indeed only exact when used to explain the margin output of the model. However, by using the `model_output` option you can use ideas from DeepSHAP to explain a non-linear transformation of the model output (this is no longer exact). Right now only a few transformations are supported, but in theory the method could work for any non-linear transform we care to support. So if there are particular transformations that you need supported I recommend posting a github issue on it, or even better asking how to make simple PR for it.


> If you compute the set of weighted key factors in margin space, you'll get a very different set of factors and weights than if you compute them in score space, which is where banks derive their top five explanations for rejecting a borrower.

It is useful here to stop and consider why you might get a different set of factors in the top 5. It is because moving the output prediction by certain amount in the margin space does not correspond to moving an equivalent amount in score space. The question about which is better really comes down to asking which is a more natural space to measure the additive effects of features. By construction the margin space of a generalized linear model is designed to be a space where additive effects make sense (that's why the logistic link function is used), and that is a good reason to consider using that space for your explanations. This is the same reason I often encourage people to think about explanations in margin space and not just use probability space. In probability space adding things together makes less sense (adding log probabilities makes sense, adding raw probabilities typically does not). I am not an expert in finance, but I suspect there could be similar issues when using score space since it, like probability, is bounded. Note that even if the bank used score space to pick a cutoff threshold, that also corresponds to a threshold in the true margin space. So my gut reaction would be to carefully consider if it would not be better to provide explanations in margin space, and just show what threshold the user needs to hit. I talked with Zest about this and it seems like it could be better to do the explanations in score space for finance, but that conclusion is not 100% clear cut.


**2) Explanation by reference**

> SHAP computes variable importance globally, which means it shows how the model behaves for every applicant (in margin space) with respect to the overall model itself. In credit risk modeling, it is often required to understand an applicant’s score in terms of another applicant or applicant population, that is, with respect to a reference population.

This is saying that you have to use the training set as the background distribution when using TreeExplainer, and cannot specify any other background dataset. This was partially true last year for TreeExplainer, but not now. Any recent version of SHAP will allow you to pass in a dataset to use as the background distribution. As noted in the original issue post, using the `data` argument to the `TreeExplainer` constructor can handle using any background distribution as the reference set. `GradientExplainer` and `DeepExplainer` also work the same way. I do however wholeheartedly agree with the main point of this section of Zest's article that it is very important to consider what reference distribution you are using. For example, there have been discussions on the SHAP github in the past that have suggested using the population of approved loans as the reference distribution when explaining a load denial. 


> Although SHAP provides fast explanations for gradient-boosted tree models, there are many other mechanisms for building scoring functions, including many alternative forms of tree models such as random forests and extremely random forests, not to mention other implementations of gradient boosting such as LightGBM.

This as well as earlier portions of the article seem to imply that shap.TreeExplainer only runs on XGBoost models. This is not true. shap.TreeExplainer supports over 20 different types of tree models, including LightGBM, scikit-learn random forests, and others. The bindings for many of these have been kindly contributed by the community, so if you find one that is not supported, you (or a friend) can add it.

**3) You want to use modeling methods other than XGBoost**

> The current implementation of SHAP cannot explain any of the other types of tree models, and cannot explain any continuous model outside a small collection, and only by importing algorithms other than SHAP.

This is unfortunately (or fortunately) not true. As noted above, TreeExplainer handles many different models out of the box, and DeepExplainer and GradientExplainer both implement algorithms that can handle deep learning models (they also both contain algorithms we designed, based on extensions to previous approaches). I don't think the algorithms in either are quite as good as TreeExplainer yet, but we (and others) are working on it :)


> What’s more, SHAP cannot explain ensembles of continuous and tree-based models, such as stacked or deeply stacked models that combine xgboost and deep neural networks."

Stacking mixed model types is an open PR request that we are still working on. You can though combine TreeExplainer and DeepExplainer today if you are just doing an ensemble of them that is some linear combination (like the mean), by just doing the same linear combination to combine the explanations.

**4) Even on a simple XGBoost model, SHAP fails to uncover the underlying geometry**

> ...the model produces nicely separated outputs. We should see that same separation when we look at the explainer outputs.

Just because you have well-separated groups of individuals in the output of the model does not mean that everyone in the same cluster of output scores should be in that output cluster for the same reason. The overlapping oval example they give is a clear example of this (in my opinion). I could be in the high risk group because I have high f1 and low f0, or because I have high f0 and low f1. Assuming f0 and f1 are meaningful input features like age and income, then these are two very different reasons for being in the same class, and so I would not expect my explanations to be the same (or similar) for everyone in that class. If we want to compare explanation methods then I would recommend using the panel of benchmarks included in the SHAP package. I should note that I talked to Zest about this and the exact reasons for the differences they found are still an open discussion. In other words, yes there is a difference, but it is not clear which one is better and for what reasons (if at some point we can run Zest's code then we could dig into this). 


**Final Note:**

It seems that Zest has worked hard on building something specific for the needs of finance, and are doing so by building something similar to GradientExplainer. I expect they have learned specific things that are important in that domain. The point Jay made about not using stuff you don't understand for important applications is of course very valid. However, given that we want to use some type of feature attribution method for finance adverse action code generation (which is a separate discussion), I am not convinced that SHAP has fundamental limitations that preclude its use in finance. For those of you that are experimenting with SHAP in this domain, I encourage you to share thoughts and notebooks.

**Comment by jabamar** [comment]
> **1) Margin Space vs Score Space**
>
> [...]
>
> > If you compute the set of weighted key factors in margin space, you'll get a very different set of factors and weights than if you compute them in score space, which is where banks derive their top five explanations for rejecting a borrower.
> 
> It is useful here to stop and consider why you might get a different set of factors in the top 5. It is because moving the output prediction by certain amount in the margin space does not correspond to moving an equivalent amount in score space. 

There is something I don't understand about this (and I think I'm not the only one with the same doubt!). I can see that the relative weights of the different factors may change after applying a non-linear transformation between the score vs margin space, but why should also the ranking change? Maybe this is a naïve point, but if the function of the transformation is continuous and monotonic, I would expect the ranking to be the same. Both the sigmoid function (to change to probability) and the cumulative function (which I understand is the transformation ZestFinance is mentioning in their article to change to margin space) are.

**Comment by slundberg** [comment]
@jabamar Good question. The reason the ranking *might* change, even with a monotonic transform, is that some features could have a large effect in rare cases and a squashing function (like sigmoid or ECDF) could take these rare cases and make them less impactful. So for example, features who's importance (when hidden) has a large effect 5% of the time and a small effect 95% of the time will shrink more from a squashing function than features that depend on moderate effects 100% of the time.

**Comment by nickpschmidt** [comment]
@jabamar, I was confused by this as well.  Beyond what is in Scott's response a couple of hours ago, it sounds like what they are talking about is just moving from probabilities to percentiles and then using a given percentile cutoff to choose to whom to make a loan.  If this is the case, then I think the issue they are bringing up is the same thing as the "explanation by reference" section of their article.  Here the issue is that most users would input the model development data into SHAP, which leads each observation's Shapley values being based on a comparison to the mean value of the model development dataset.  But if the bank is choosing acceptance based on a given percentile, it is pretty unlikely that the chosen percentile would be associated with the average default probability in the development data.  If that is the case, then it is true that this SHAP output will likely not give the correct results.  But, from what @slundberg says, it sounds like using a different reference distribution would correct for this issue.  We discussed this in another issue in this repository if you are interested in reading more.
 
I was also surprised by their point about a bank making a decision based on a given percentile.  While I do see this done in marketing, where a marketing budget dictates the depth of the offer, I have never seen this in a credit model (that certainly does not mean that it is not done).  It would surprise me if a bank chose a percentile cutoff without the decision fundamentally being based on the both the underlying probability of default and the distribution of the predictions from the model.  Otherwise they would not be considering the NPV and Value-at-Risk of the loan portfolio.  For example, giving loans to 20% of a subprime population will lead to very different financial results than giving loans to 20% of a prime population.  Given this, while the cutoff chosen may be communicated to non-modelers as being, "we make offers to 20% of applicants," the choice of 20% is going to be based on the underlying predictions and distributions associated with that 20%. To me, that means that the probability (or log-odds) is the driver of decisions and should be the basis for explanations.
 
Finally, just for people who are less familiar with the "score space" idea, here are a couple of points: the problem that I believe they are pointing out is that when you output to the probability space most of the estimates are lumped together.  As a very hypothetical example, if the average probability of default is 10%, we might see 80% of observations between 9.5% and 10.5%.  People get confused by this, so a transformation is applied which spreads out the distribution.  The article pointed out a shift to percentiles; two other transformations I have frequently seen are transformations either to a range that looks like the distribution for a credit score (e.g., FICO or Vantage), or to a distribution that makes the outcomes approximately normally distributed (those may be similar - I don't remember what the distribution of credit scores looks like).  I think that those transformations are generally accomplished through the CDF of the margin space, so they are effectively the same as changes to percentiles.

I do hope that Zest clarifies some of the points because I know them to be brilliant, and they are doing very cool stuff that I highly respect.  So I would really like to learn more from them about their points.

**Comment by InfiniteDoLoop** [comment]
@slundberg, first off congrats on completing your PhD! And thank you for providing a lot of clarification on the article. I appreciate that you discussed the article content further with Zest and shared your thoughts with us.

I had a follow up question on **2) Explanation by reference**

So the acceptance/rejection mechanism seems to be based on defining a score cutoff. Let's say the rule is that if an application has a probability of default > X%, then reject it.
But in Zest's framework, it seems to focus on comparing each application to a background dataset of approved applications. The issue with this is that, most likely, E[Score] < X% and so you're not really comparing the applicant's score to the true cutoff, but to the mean score of the approved population.

It seems like one way to overcome this issue would require some sampling strategy where the background dataset fed into the explainer would need to have E[Score] = X%. This is definitely not straightforward, but just one idea. Open to any other suggestions you may have on addressing such an issue.

@jabamar, I think what Scott is trying to describe is following when moving from the unbounded log odds space to bounded (non-linear) probability space:

![shap](https://user-images.githubusercontent.com/40615060/60366877-c8013580-99ba-11e9-8edb-5c0283b8466a.png)

On the X-axis - you have the log odds, y-axis - probability.
Note that even though we moved on the log odds space by equal units, the movement in probability space is very different. That's why Scott mentioned a squashing function (like Sigmoid) would lessen the impact of rare cases on the final probability outcome.

**Comment by nickpschmidt** [comment]
@InfiniteDoLoop, isn't your example a result of ordering and not taking into account the averaging idea of Shapley Values?  As I understand it, the idea is, "What is the average marginal effect of the inclusion of a particular feature over any ordering of inclusion?"  So, I would think that if you want to map the idea of Shapley Values from log-odds to probability space, you have to first calculate the effect of feature 1 and then feature 2; calculate the marginal effects (like you did above), then do the opposite ordering: assume feature 2 is the first feature, then feature 1.  This gives equal but opposite effects.  Averaging these two together, you get the result that each feature contributed the same amount (in your example it would be (38%+10%)/2 =24%).
Maybe it's not that simple?  It might be that, in order to capture the true average marginal effect, you have to run the whole Shapley process at one time, including both running through the trees as well as applying the probability transformation.  Just casually thinking about it, I don't think that it is necessary, but it seems possible that the transformation function does something that messes up the final Shapley Values if you don't do everything at one time.

**Comment by InfiniteDoLoop** [comment]
@nickpschmidt, I wasn't particularly trying to highlight any idea of ordering in that example. You could assume that these are the final SHAP/Shapley values.
Rather was trying to highlight that the rate of change in probability with respect to log odds is very volatile. This is why a large movement in one of the extreme ends of log odds space can appear to look rather small in probability space.

**Comment by nickpschmidt** [comment]
@InfiniteDoLoop, I think I missed your point then, thanks.  You are (clearly) right that a simple transformation of log-odds Shapleys to probabilities can be inconsistent.  But my point is that it appears that the ranking of Shapley Values can survive the transformation to probability space as long as you continue permuting the values when you transform them into probabilities.

**Comment by slundberg** [comment]
@InfiniteDoLoop Good question about the reference distribution. I believe you are right they are using the set of approved applicants as the reference distribution. It would be interesting to take a sample of past applicants right at the cut-off boundary and use those as the background reference distribution. This would make your feature importances relative to this group. So being better than them would mean "acceptance" and worse would mean "rejection". It is nice because it makes "zero" mean something important, so positive and negative values would connect directly to being above or below the acceptance threshold. In other words then if all your weights summed to a positive number then you will be accepted :)

@jabamar As a quick followup clarification I posted an example notebook showing how feature importance orders can be changed by using probability space instead of log-odds margin space: https://github.com/slundberg/shap/blob/master/notebooks/kernel_explainer/Squashing%20Effect.ipynb

**Comment by jlb0666** [comment]
Hi there -- we at [ZestFinance](http://zestfinance.com/) appreciate
the contributions @slundberg and the community have made and
all the time Scott has spent with us discussing ML model
explainability. Scott is a huge talent and wherever he goes after UW
will be lucky to have him. Scott's notebook nicely illustrates one of our points.  We also posted a follow-up [here](https://www.zestfinance.com/blog/part-deux-why-not-just-use-shap).  Alas,
we did miss some of the more recent SHAP functionality in our initial
analysis (our internal version of the package was pinned to a release from last year), but we do still have some concerns with the strong assumptions made by the core SHAP algorithms when moving between margin space and score space as is required in many financial services
applications.  Namely that: (1) variables are assumed to be statistically independent and (2)
data can be filled in assuming that missing values are missing
completely at random.  We are preparing a manuscript describing a
method that overcomes these issues and will be sure to follow up here when it's published.  
-Jay Budzik, CTO, ZestFinance

**Comment by JingwuChen** [comment]
> @InfiniteDoLoop Good question about the reference distribution. I believe you are right they are using the set of approved applicants as the reference distribution. It would be interesting to take a sample of past applicants right at the cut-off boundary and use those as the background reference distribution. This would make your feature importances relative to this group. So being better than them would mean "acceptance" and worse would mean "rejection". It is nice because it makes "zero" mean something important, so positive and negative values would connect directly to being above or below the acceptance threshold. In other words then if all your weights summed to a positive number then you will be accepted :)
> 
> @jabamar As a quick followup clarification I posted an example notebook showing how feature importance orders can be changed by using probability space instead of log-odds margin space: https://github.com/slundberg/shap/blob/master/notebooks/kernel_explainer/Squashing%20Effect.ipynb

hello,thank you for reply,but the link is invalid

**Comment by ehsanfathi77** [comment]
@jlb0666 It would be great if you could kindly share your promised follow up manuscript.
@JingwuChen you can find the notebook here: https://github.com/shap/shap/blob/master/notebooks/tabular_examples/model_agnostic/Squashing%20Effect.ipynb

**Comment by eddyfathi** [comment]
@slundberg since it's been a while since 2019, do you have any new insight to share about this issue?

**Comment by ehsanfathi77** [comment]
@slundberg  since it's been a while since 2019, do you have any new insight to share about this issue?
